modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/schema/IgniteTable.java

```diff
@@ -97,18 +116,78 @@ public String name() {
         return statistic;
     }
 
+    /** */
+    public TableDescriptor descriptor() {
+        return desc;
+    }
+
     /** {@inheritDoc} */
     @Override public RelNode toRel(RelOptTable.ToRelContext context, RelOptTable relOptTable) {
         RelOptCluster cluster = context.getCluster();
+
+        return toRel(cluster, relOptTable, PK_INDEX_NAME);
+    }
+
+    /**
+     * Converts table into relational expression.
+     *
+     * @param cluster Custer.
+     * @param relOptTable Table.
+     * @return Table relational expression.
+     */
+    public IgniteTableScan toRel(RelOptCluster cluster, RelOptTable relOptTable, String idxName) {
         RelTraitSet traitSet = cluster.traitSetOf(IgniteConvention.INSTANCE)
-            .replaceIfs(RelCollationTraitDef.INSTANCE, this::collations)
             .replaceIf(DistributionTraitDef.INSTANCE, this::distribution);
 
-        return new IgniteTableScan(cluster, traitSet, relOptTable);
+        IgniteIndex idx = indexes.get(idxName);
+
+        if (idx == null)
+            return null;
+
+        traitSet = traitSet.replaceIf(RelCollationTraitDef.INSTANCE, idx::collation);
+
+        return new IgniteTableScan(cluster, traitSet, relOptTable, idxName,  null);
     }
 
-    /** {@inheritDoc} */
-    @Override public NodesMapping mapping(PlanningContext ctx) {
+    /**
+     * @return Indexes for the current table.
+     */
+    public Map<String, IgniteIndex> indexes() {
```

Comment:
Lets introduce another interface like IndexedTable and put related methods into it, in future it will help us to use mocks in tests, another point is that we are going to rethink how we manage services and components and put everything into an IoC container. Right interfaces and abstractions are necessary.